src/Servers/IIS/src/AspNetCoreModuleV2/RequestHandlerLib/environmentvariablehelpers.h

```diff
@@ -24,422 +26,77 @@ class ENVIRONMENT_VAR_HELPERS
-            // As UI and CMD does not allow empty value, ignore this environment var
-            if (dwError != ERROR_ENVVAR_NOT_FOUND && dwError != ERROR_SUCCESS)
-            {
-                hr = HRESULT_FROM_WIN32(dwError);
-                goto Finished;
-            }
-        }
-        else if (dwResult > strStartupAssemblyEnv.QuerySizeCCH())
-        {
-            // have to increase the buffer and try get environment var again
-            strStartupAssemblyEnv.Reset();
-            strStartupAssemblyEnv.Resize(dwResult + (DWORD)wcslen(HOSTING_STARTUP_ASSEMBLIES_VALUE) + 1);
-            dwResult = GetEnvironmentVariable(HOSTING_STARTUP_ASSEMBLIES_ENV_STR,
-                strStartupAssemblyEnv.QueryStr(),
-                strStartupAssemblyEnv.QuerySizeCCH());
-            if (dwResult <= 0)
+            if (hostingStartupValues.find(HOSTING_STARTUP_ASSEMBLIES_ENV_STR) == std::wstring::npos)
             {
-                hr = E_UNEXPECTED;
-                goto Finished;
+                hostingStartupValues += std::wstring(L";") + HOSTING_STARTUP_ASSEMBLIES_VALUE;
             }
-            fFound = TRUE;
-        }
-        else
-        {
-            fFound = TRUE;
-        }
-
-        strStartupAssemblyEnv.SyncWithBuffer();
-        if (strStartupAssemblyEnv.IndexOf(HOSTING_STARTUP_ASSEMBLIES_VALUE) == -1)
-        {
-        if (fFound)
-        {
-            strStartupAssemblyEnv.Append(L";");
-        }
-        strStartupAssemblyEnv.Append(HOSTING_STARTUP_ASSEMBLIES_VALUE);
-        }
-
-        // the environment variable was not defined, create it and add to hashtable
-        pHostingEntry = new ENVIRONMENT_VAR_ENTRY();
 
-        if (FAILED(hr = pHostingEntry->Initialize(HOSTING_STARTUP_ASSEMBLIES_NAME, strStartupAssemblyEnv.QueryStr())) ||
-            FAILED(hr = pEnvironmentVarTable->InsertRecord(pHostingEntry)))
-        {
-            goto Finished;
+            environmentVariables.insert_or_assign(HOSTING_STARTUP_ASSEMBLIES_ENV_STR, hostingStartupValues);
         }
 
-    Skipped:
-        *ppEnvironmentVarTable = pEnvironmentVarTable;
-        pEnvironmentVarTable = NULL;
-
-    Finished:
-        if (pHostingEntry != NULL)
+        for (auto& environmentVariable : environmentVariables)
         {
-            pHostingEntry->Dereference();
-            pHostingEntry = NULL;
+            environmentVariable.second = Environment::ExpandEnvironmentVariables(environmentVariable.second);
         }
 
-        if (pIISAuthEntry != NULL)
-        {
-            pIISAuthEntry->Dereference();
-            pIISAuthEntry = NULL;
-        }
-
-        if (pEnvironmentVarTable != NULL)
-        {
-            pEnvironmentVarTable->Clear();
-            delete pEnvironmentVarTable;
-            pEnvironmentVarTable = NULL;
-        }
-        return hr;
+        return environmentVariables;
     }
 
     static
-    HRESULT
+    std::map<std::wstring, std::wstring, ignore_case_comparer>
```

Comment:
Why return the environment variable map when you already using a ref for the parameter?

Reply:
It's a modified copy of the original map. It takes original as a const reference to avoid one extra copy.

Reply:
Why not just modify the map passed in to avoid the copy?

Reply:
Because that's the original one stored in options and in OutOfProc InitEnvironmentVariablesTable might be called multiple times.